### PR TITLE
Add new API in SBTarget for loading core from SBFile

### PR DESCRIPTION
**Pull request by GeorgeHuyubo**
Add a new API in SBTarget to Load Core from a SBFile.
This will enable a target to load core from a file descriptor.
So that in coredumper, we don't need to write core file to disk, instead we can pass the input file descriptor to lldb directly.

Test:
```
(lldb) script
Python Interactive Interpreter. To exit, type 'quit()', 'exit()' or Ctrl-D.
>>> file_object = open("/home/hyubo/210hda79ms32sr0h", "r")
>>> fd=file_object.fileno()
>>> file = lldb.SBFile(fd,'r', True)
>>> error = lldb.SBError()
>>> target = lldb.debugger.CreateTarget(None)
>>> target.LoadCore(file,error)
SBProcess: pid = 56415, state = stopped, threads = 1
```

### CODE REVIEW

File: lldb/include/lldb/API/SBTarget.h [hunk 1]
```diff
@@ -184,6 +184,8 @@ class LLDB_API SBTarget {
 
   SBProcess LoadCore(const char *core_file);
   SBProcess LoadCore(const char *core_file, lldb::SBError &error);
+  SBProcess LoadCore(SBFile &file);
```
Comment (JDevlieghere) [inline lldb/include/lldb/API/SBTarget.h:187]: Did you add this for consistency? I don't think we should be adding an overload without error handling.

File: lldb/include/lldb/Target/PostMortemProcess.h [hunk 2]
```diff
@@ -24,7 +26,20 @@ class PostMortemProcess : public Process {
...
+  PostMortemProcess(lldb::TargetSP target_sp, lldb::ListenerSP listener_sp,
+                    lldb::FileSP file_sp)
+      : Process(target_sp, listener_sp), m_core_file(file_sp) {}
+
   bool IsLiveDebugSession() const override { return false; }
+
+  FileSpec GetCoreFile() const override {
+    const FileSpec file_spec;
+    m_core_file->GetFileSpec(const_cast<FileSpec &>(file_spec));
+    return file_spec;
```
Comment (JDevlieghere) [inline lldb/include/lldb/Target/PostMortemProcess.h:38]: I'm not sure I understand what's going on here. You're marking the local variable as `const` only to immediately `const_cast` it away in the line below? Can't you avoid all this like so:

```
FileSpec file_spec;
m_core_file->GetFileSpec(file_spec);
```

or even better, why doesn't `GetFileSpec` return a `FileSpec` instead of an out parameter?
```

File: lldb/include/lldb/Target/Process.h [hunk 3]
```diff
@@ -354,12 +355,10 @@ class Process : public std::enable_shared_from_this<Process>,
...
-  static constexpr int g_all_event_bits = eBroadcastBitStateChanged 
-                                        | eBroadcastBitInterrupt
-                                        | eBroadcastBitSTDOUT 
-                                        | eBroadcastBitSTDERR
-                                        | eBroadcastBitProfileData 
-                                        | eBroadcastBitStructuredData;
+  static constexpr int g_all_event_bits =
+      eBroadcastBitStateChanged | eBroadcastBitInterrupt | eBroadcastBitSTDOUT |
+      eBroadcastBitSTDERR | eBroadcastBitProfileData |
+      eBroadcastBitStructuredData;
```
Comment (JDevlieghere) [inline lldb/include/lldb/Target/Process.h:361]: Unrelated formatting change?

File: lldb/source/Commands/CommandObjectTarget.cpp [hunk 4]
```diff
@@ -2104,9 +2113,10 @@ class CommandObjectTargetModulesDumpSymtab
                 result.GetOutputStream().EOL();
                 result.GetOutputStream().EOL();
               }
-              if (INTERRUPT_REQUESTED(GetDebugger(), 
-                    "Interrupted in dump symtab list with {0} of {1} dumped.", 
-                    num_dumped, num_matches))
+              if (INTERRUPT_REQUESTED(
+                      GetDebugger(),
+                      "Interrupted in dump symtab list with {0} of {1} dumped.",
+                      num_dumped, num_matches))
```
Comment (JDevlieghere) [inline lldb/source/Commands/CommandObjectTarget.cpp:2119]: Please upstage unrelated code changes. I suspect this is the result of our editor removing trailing whitespace (I do the same) and then clang-format considers this a modified line and formats it.

**Review by JDevlieghere — CHANGES_REQUESTED** [review]
I left some comments inline.

File: lldb/include/lldb/API/SBTarget.h (hunk 1 above)
Comment (bulbazord) [reply lldb/include/lldb/API/SBTarget.h:187]: +1, please don't add a new API that gives no feedback about what may have gone wrong.

File: lldb/include/lldb/Target/PostMortemProcess.h (hunk 2 above)
Comment (bulbazord) [reply lldb/include/lldb/Target/PostMortemProcess.h:38]: +1

File: lldb/include/lldb/Target/Process.h [hunk 5]
```diff
@@ -390,12 +389,12 @@ class Process : public std::enable_shared_from_this<Process>,
...
-/// for changes in a process's lifetime.
-///
-/// \see RegisterNotificationCallbacks (const Notifications&) @see
-/// UnregisterNotificationCallbacks (const Notifications&)
+
+  /// A notification structure that can be used by clients to listen
+  /// for changes in a process's lifetime.
+  ///
+  /// \see RegisterNotificationCallbacks (const Notifications&) @see
+  /// UnregisterNotificationCallbacks (const Notifications&)
```
Comment (bulbazord) [inline lldb/include/lldb/Target/Process.h:397]: Unrelated formatting change

File: lldb/include/lldb/Target/Process.h [hunk 6]
```diff
@@ -1469,6 +1468,14 @@ class Process : public std::enable_shared_from_this<Process>,
 
   virtual bool IsLiveDebugSession() const { return true; };
 
+
+  /// Provide a way to retrieve the core dump file that is loaded for debugging.
+  /// Only available if IsLiveDebugSession() returns true.
```
Comment (bulbazord) [inline lldb/include/lldb/Target/Process.h:1473]: Shouldn't it only be available if `IsLiveDebugSession` is false? You're debugging a core dump which isn't really a "live debug session" since you're examining the corpse of a process, no?

File: lldb/source/API/SBTarget.cpp [hunk 7]
```diff
@@ -244,9 +245,45 @@ SBProcess SBTarget::LoadCore(const char *core_file, lldb::SBError &error) {
...
+    if (process_sp) {
+      error.SetError(process_sp->LoadCore());
+      if (error.Success())
+        sb_process.SetSP(process_sp);
+    } else {
+      error.SetErrorString("Failed to create the process");
+    }
+  } else {
+    error.SetErrorString("SBTarget is invalid");
+  }
```
Comment (bulbazord) [inline lldb/source/API/SBTarget.cpp:267]: If you flip the condition that checks target_sp for validity, you can save a level of indentation for readability (if that matters to you).

File: lldb/source/Plugins/Process/FreeBSDKernel/ProcessFreeBSDKernel.cpp [hunk 8]
```diff
@@ -67,30 +67,33 @@ class ProcessFreeBSDKernelKVM : public ProcessFreeBSDKernel {
...
 
 lldb::ProcessSP ProcessFreeBSDKernel::CreateInstance(lldb::TargetSP target_sp,
                                                      ListenerSP listener_sp,
-                                                     const FileSpec *crash_file,
+                                                     lldb::FileSP crash_file_sp,
                                                      bool can_connect) {
   ModuleSP executable = target_sp->GetExecutableModule();
   if (crash_file && !can_connect && executable) {
+    const FileSpec file_spec;
+    crash_file_sp->GetFileSpec(const_cast<FileSpec &>(file_spec));
```
Comment (bulbazord) [inline lldb/source/Plugins/Process/FreeBSDKernel/ProcessFreeBSDKernel.cpp:81]: Same here as before, why are you making a const file spec and then const casting to fill it in?

File: lldb/source/Plugins/Process/MacOSX-Kernel/ProcessKDP.cpp [hunk 9]
```diff
@@ -102,10 +102,10 @@ void ProcessKDP::Terminate() {
 
 lldb::ProcessSP ProcessKDP::CreateInstance(TargetSP target_sp,
                                            ListenerSP listener_sp,
-                                           const FileSpec *crash_file_path,
+                                           FileSP crash_file_sp,
                                            bool can_connect) {
   lldb::ProcessSP process_sp;
-  if (crash_file_path == NULL)
+  if (crash_file_sp == NULL)
```
Comment (bulbazord) [inline lldb/source/Plugins/Process/MacOSX-Kernel/ProcessKDP.cpp:108]: Since you're already here, you can drop the `== NULL` and make it a little nicer.

File: lldb/source/Plugins/Process/elf-core/ProcessElfCore.cpp [hunk 10]
```diff
@@ -49,17 +50,18 @@ void ProcessElfCore::Terminate() {
...
+  if (crash_file_sp && !can_connect) {
     // Read enough data for an ELF32 header or ELF64 header Note: Here we care
     // about e_type field only, so it is safe to ignore possible presence of
     // the header extension.
     const size_t header_size = sizeof(llvm::ELF::Elf64_Ehdr);
-
-    auto data_sp = FileSystem::Instance().CreateDataBuffer(
-        crash_file->GetPath(), header_size, 0);
+    const FileSpec file_spec;
+    crash_file_sp->GetFileSpec(const_cast<FileSpec &>(file_spec));
```
Comment (bulbazord) [inline lldb/source/Plugins/Process/elf-core/ProcessElfCore.cpp:62]: Here too

File: lldb/source/Plugins/Process/gdb-remote/ProcessGDBRemote.cpp [hunk 11]
```diff
@@ -196,9 +196,10 @@ void ProcessGDBRemote::Terminate() {
 
 lldb::ProcessSP ProcessGDBRemote::CreateInstance(
     lldb::TargetSP target_sp, ListenerSP listener_sp,
-    const FileSpec *crash_file_path, bool can_connect) {
+                                                 lldb::FileSP crash_file_sp,
+                                                 bool can_connect) {
   lldb::ProcessSP process_sp;
-  if (crash_file_path == nullptr)
+  if (crash_file_sp == nullptr)
```
Comment (bulbazord) [inline lldb/source/Plugins/Process/gdb-remote/ProcessGDBRemote.cpp:202]: You can also drop the `== nullptr` here too

File: lldb/source/Plugins/Process/mach-core/ProcessMachCore.cpp [hunk 12]
```diff
@@ -61,13 +62,15 @@ void ProcessMachCore::Terminate() {
...
+                                                lldb::FileSP crash_file_sp,
                                                 bool can_connect) {
   lldb::ProcessSP process_sp;
-  if (crash_file && !can_connect) {
+  if (crash_file_sp && !can_connect) {
     const size_t header_size = sizeof(llvm::MachO::mach_header);
-    auto data_sp = FileSystem::Instance().CreateDataBuffer(
-        crash_file->GetPath(), header_size, 0);
+    const FileSpec file_spec;
+    crash_file_sp->GetFileSpec(const_cast<FileSpec &>(file_spec));
```
Comment (bulbazord) [inline lldb/source/Plugins/Process/mach-core/ProcessMachCore.cpp:71]: Also here

File: lldb/include/lldb/API/SBTarget.h (hunk 1 above)
Comment (clayborg) [reply lldb/include/lldb/API/SBTarget.h:187]: Just add the error to your function call:
```
SBProcess LoadCore(SBFile &file, lldb::SBError &error);
```
The first `SBProcess LoadCore(const char *core_file);` should be marked as deprecated since we want people to use `SBProcess LoadCore(const char *core_file, lldb::SBError &error);` so they get error information back. In our public API we never remove an older API, and that is why `SBProcess LoadCore(const char *core_file);` is still there.

File: lldb/include/lldb/Target/Target.h [hunk 13]
```diff
@@ -627,7 +628,7 @@ class Target : public std::enable_shared_from_this<Target>,
   // used.
   const lldb::ProcessSP &CreateProcess(lldb::ListenerSP listener_sp,
                                        llvm::StringRef plugin_name,
-                                       const FileSpec *crash_file,
+                                       lldb::FileSP crash_file,
                                        bool can_connect);
```
Comment (clayborg) [inline lldb/include/lldb/Target/Target.h:632]: I think we can have two interfaces in Target.h: 
- the previous one that taks a "const FileSpec *" 
- the new one that takes a lldb::FileSP
We have a 4 locations that have to manually create a file the code:
```
auto file = FileSystem::Instance().Open(
        filespec, lldb_private::File::eOpenOptionReadOnly);
...
```
This will also help reduce the number of changes in this patch. Then the version that takes a "const FileSpec *" can do the `FileSystem::Instance().Open(...)` and just call the other overload that takes a `lldb::FileSP`

File: lldb/source/API/SBTarget.cpp [hunk 14]
```diff
@@ -244,9 +245,38 @@ SBProcess SBTarget::LoadCore(const char *core_file, lldb::SBError &error) {
...
```
Comment (clayborg) [inline lldb/source/API/SBTarget.cpp:267]: If we leave the Target::CreateProcess overload as suggested above, this entire change goes away.

File: lldb/source/Commands/CommandObjectTarget.cpp [hunk 15]
```diff
@@ -427,8 +427,17 @@ class CommandObjectTargetCreate : public CommandObjectParsed {
...
+        if (!file) {
+          result.AppendErrorWithFormatv(
+              "Failed to open the core file '{0}': '{1}.'\n",
+              core_file.GetPath(), llvm::toString(file.takeError()));
+        }
+
         ProcessSP process_sp(target_sp->CreateProcess(
-            GetDebugger().GetListener(), llvm::StringRef(), &core_file, false));
+            GetDebugger().GetListener(), llvm::StringRef(),
+            std::move(file.get()), false));
```
Comment (clayborg) [inline lldb/source/Commands/CommandObjectTarget.cpp:440]: All of this goes away if we leave the original Target::CreateProcess overload as mentioned above.

File: lldb/source/Core/IOHandlerCursesGUI.cpp [hunk 16]
```diff
@@ -3174,8 +3174,17 @@ class TargetCreateFormDelegate : public FormDelegate {
...
+        core_file_spec, lldb_private::File::eOpenOptionReadOnly);
+
+    if (!core_file) {
+      SetError(llvm::toString(core_file.takeError()).c_str());
+      return;
+    }
+
+    ProcessSP process_sp(
+        target_sp->CreateProcess(m_debugger.GetListener(), llvm::StringRef(),
+                                 std::move(core_file.get()), false));
```
Comment (clayborg) [inline lldb/source/Core/IOHandlerCursesGUI.cpp:3187]: All of this goes away if we leave the original Target::CreateProcess overload as mentioned above.

File: lldb/source/Plugins/Process/gdb-remote/ProcessGDBRemote.cpp [hunk 17]
```diff
@@ -194,11 +194,12 @@ void ProcessGDBRemote::Terminate() {
...
-lldb::ProcessSP ProcessGDBRemote::CreateInstance(
-    lldb::TargetSP target_sp, ListenerSP listener_sp,
-    const FileSpec *crash_file_path, bool can_connect) {
+lldb::ProcessSP ProcessGDBRemote::CreateInstance(lldb::TargetSP target_sp,
+                                                 ListenerSP listener_sp,
+                                                 lldb::FileSP crash_file_sp,
+                                                 bool can_connect) {
   lldb::ProcessSP process_sp;
-  if (crash_file_path == nullptr)
+  if (crash_file_sp)
```
Comment (clayborg) [inline lldb/source/Plugins/Process/gdb-remote/ProcessGDBRemote.cpp:202]: The logic is inverted from what it used to be, this should be:
```
if (!crash_file_sp)
```

File: lldb/source/Plugins/Process/MacOSX-Kernel/ProcessKDP.cpp [hunk 18]
```diff
@@ -102,10 +102,10 @@ void ProcessKDP::Terminate() {
 
 lldb::ProcessSP ProcessKDP::CreateInstance(TargetSP target_sp,
                                            ListenerSP listener_sp,
-                                           const FileSpec *crash_file_path,
+                                           FileSP crash_file_sp,
                                            bool can_connect) {
   lldb::ProcessSP process_sp;
-  if (crash_file_path == NULL)
+  if (crash_file_sp)
```
Comment (clayborg) [inline lldb/source/Plugins/Process/MacOSX-Kernel/ProcessKDP.cpp:108]: The logic is inverted from what it used to be, this should be:
```
if (!crash_file_sp)
```

**Review by clayborg — CHANGES_REQUESTED** [review]

File: lldb/include/lldb/Target/Target.h (hunk 13 above)
Comment (GeorgeHuyubo) [reply lldb/include/lldb/Target/Target.h:632]: If we move the open file code into Target.cpp, it's hard to handle error. Now in these 4 locations we have slightly different ways to report the error when the file is failed to open.
```
        if (!file) {
          result.AppendErrorWithFormatv(
              "Failed to open the core file '{0}': '{1}.'\n",
              core_file.GetPath(), llvm::toString(file.takeError()));
        }
```

**Review by bulbazord — APPROVED** [review]
LGTM I think, but wait a little bit for Jonas and/or Greg to give it another pass.

File: lldb/include/lldb/API/SBTarget.h [hunk 19]
```diff
@@ -184,6 +184,7 @@ class LLDB_API SBTarget {
 
   SBProcess LoadCore(const char *core_file);
   SBProcess LoadCore(const char *core_file, lldb::SBError &error);
+  SBProcess LoadCore(SBFile &file, lldb::SBError &error);
```
Comment (bulbazord) [inline lldb/include/lldb/API/SBTarget.h:187]: Actually, can the `SBFile &` be marked `const`? This shouldn't modify the SBFile if I'm reading this correctly.

File: lldb/source/API/SBTarget.cpp [hunk 20]
```diff
@@ -260,6 +261,31 @@ SBProcess SBTarget::LoadCore(const char *core_file, lldb::SBError &error) {
...
+  return sb_process;
+}
```
Comment (JDevlieghere) [inline lldb/source/API/SBTarget.cpp:287]: This could be a lot simpler and easier to read with early returns, with the added advantage that the error message is closer to the condition it corresponds to. The function below is a good example of that.

**Comment by clayborg** [comment]
I am not sure how this actually works? We pass in a SBFile and then get the FileSpec from it and then somehow that allows us to open a file that doesn't actually exist on disk? Or does this work _because_ the `fd` we had is actually backed by some file on disk? In that case we didn't even need this new API as we could have extracted the file path from the descriptor and just passed that into LLDB.

The story is that linux allows a tool to get run when a core file is created and it gets a integer file descriptor for the core file that is used to access the core file. Prior to this fix we would write the file to disk and then load it into the debugger. Now we want to use the "fd" to load the core file, so our SBFile must be able to extract the actual backing file and then we will need to re-open this file in lldb???

**Comment by omjavaid** [comment]
I have reverted this temporarily as it broke  TestLinuxCore.py on lldb-*-windows. Kindly have a look at
    https://lab.llvm.org/buildbot/#/builders/219/builds/7014
